src/ext.rs

```diff
@@ -115,8 +117,54 @@ impl HeaderCaseMap {
 
     pub(crate) fn append<N>(&mut self, name: N, orig: Bytes)
     where
-        N: IntoHeaderName,
+        N: IntoHeaderName + std::fmt::Display,
     {
         self.0.append(name, orig);
     }
 }
+
+#[derive(Clone, Debug)]
+/// Hashmap<Headername, numheaders with that name>
+pub(crate) struct OriginalHeaderOrder(HashMap<HeaderName, usize>, Vec<(HeaderName, usize)>);
```

Comment:
I'm immediately understanding the separate hashmap and vector. Do they need to be separate? Can the hashmap just contain the vector as the value, instead of the number?

Reply:
I'm just using the hash map to keep track off how many headers of the same name have already been seen. This makes the bookkeeping around inserting and appending easier.
Side Note: if the number of header names is low, it might be faster to use a BTreeMap here to avoid string hashing.

The vector is what really stores the ordering of the headers. ex: `vec[i] = (headerName, idx)`, This vector is ordered such that the `i`th element represents the `i`th header that came in off the line. The `headerName` and `idx` are then used elsewhere to index into the multi map that stores the header values.
```rust
           for (name, idx) in headers.orig_order.get_in_order() {
            let orig_name = headers.orig_casing.get_all(&name).nth(*idx).unwrap();
            let value = headers.headers.get_all(name).iter().nth(*idx).unwrap();
```

Reply:
Hm, ok. What do you think of giving the fields names, and adding a comment above each field explaining how they relate? That should help Future Us make changes some day.